### PR TITLE
Support for syncing on mergemock

### PR DESCRIPTION
**Pull request by Giulio2002**
## Things Added
* Code for Block Proposing
* Validators can now be ran on Erigon
* Private testnet with proof-of-stake enabled can be ran on Erigon
* `mergemock` can be ran on Erigon
* All test vectors from EF pass successfully

## Prerequisite

* Erigon-lib: https://github.com/ledgerwatch/erigon-lib/pull/217
* Interfaces : https://github.com/ledgerwatch/interfaces/pull/87

Above PRs must be merged for this PR to not be a draft @AskAlexSharov 

## Things missing before moving to M2 Kinstugi Milestones:

* Disable P2P Gossip accordingly to spec
*  EIP-2124

### CODE REVIEW

**Comment by Giulio2002** [comment]
More things to add: if ran against `mergemock` VerifyHeader could fail, probably a bug in `mergemock`, for now i commented the `verifyHeader` statement. other thing to say, if we are on the tip of the chain and receive the new head, i made it so we process the bodies in stage headers, i will change it in a next PR, this one is too big already honestly.

File: .gitignore [hunk 1]
```diff
@@ -16,7 +16,7 @@
 build/_vendor/pkg
 /*.a
 docs/readthedocs/build
-
+genesis.json
```
Comment (yperbasis) [inline .gitignore:19]: Hmm. Is it OK to add `genesis.json` to `.gitignore`?

Comment (Giulio2002) [reply .gitignore:19]: the `.gitignore` will be cleaned prior to merging, i put it like that so that i dont have to make a new one every time

File: eth/stagedsync/stage_mining_finish.go [hunk 2]
```diff
@@ -5,17 +5,19 @@ import (
...
-	chainConfig params.ChainConfig
-	engine      consensus.Engine
-	sealCancel  <-chan struct{}
-	miningState MiningState
+	db             kv.RwDB
+	chainConfig    params.ChainConfig
+	engine         consensus.Engine
+	sealCancel     <-chan struct{}
+	miningState    MiningState
+	assembledBlock *types.Block
```
Comment (yperbasis) [inline eth/stagedsync/stage_mining_finish.go:20]: Why do you need field? I see that it's only written to, but never read from.

**Comment by Giulio2002** [comment]
We use it to communicate the mined block to ForkchoiceUpdated, so that we keep mining separate from ethbackend

**Comment by yperbasis** [comment]
> We use it to communicate the mined block to ForkchoiceUpdated, so that we keep mining separate from ethbackend

But where's the code that reads `MiningFinishCfg.assembledBlock`?

**Comment by Giulio2002** [comment]
> 

here https://github.com/ledgerwatch/erigon/blob/45cfc7ed83aafd4d55f4a0c728fed0674aa5ae79/ethdb/privateapi/ethbackend.go#L308 `s.assembledBlock` is the same pointer as in the mining staged sync

**Comment by yperbasis** [comment]
> https://github.com/ledgerwatch/erigon/blob/45cfc7ed83aafd4d55f4a0c728fed0674aa5ae79/ethdb/privateapi/ethbackend.go#L308
> 
> `s.assembledBlock` is the same pointer as in the mining staged sync

Hmm. I'm afraid I'm still not following. Where is the code that ensures that `MiningFinishCfg.assembledBlock` is the same as `EthBackendServer.assembledBlock`?

**Comment by Giulio2002** [comment]
> > https://github.com/ledgerwatch/erigon/blob/45cfc7ed83aafd4d55f4a0c728fed0674aa5ae79/ethdb/privateapi/ethbackend.go#L308
> > 
> > `s.assembledBlock` is the same pointer as in the mining staged sync
> 
> Hmm. I'm afraid I'm still not following. Where is the code that ensures that `MiningFinishCfg.assembledBlock` is the same as `EthBackendServer.assembledBlock`?

https://github.com/ledgerwatch/erigon/blob/45cfc7ed83aafd4d55f4a0c728fed0674aa5ae79/eth/backend.go#L138

https://github.com/ledgerwatch/erigon/blob/45cfc7ed83aafd4d55f4a0c728fed0674aa5ae79/eth/backend.go#L402

https://github.com/ledgerwatch/erigon/blob/45cfc7ed83aafd4d55f4a0c728fed0674aa5ae79/eth/backend.go#L410

It is the same pointer

**Comment by yperbasis** [comment]
> It is the same pointer

Ah, OK, I see. The problem is that you need a syncronization mechanism to prevent data races since the data will be accessed from multiple goroutines. You should use something like `atomic.Value` (or a channel or a mutex perhaps).

**Comment by yperbasis** [comment]
Something like ["Concurrency in Go" by Katherine Cox–buday](https://www.amazon.co.uk/Concurrency-Go-Katherine-Cox-buday/dp/1491941197) is a nice read :)

**Comment by Giulio2002** [comment]
> Something like ["Concurrency in Go" by Katherine Cox–buday](https://www.amazon.co.uk/Concurrency-Go-Katherine-Cox-buday/dp/1491941197) is a nice read :)

Yeah, I do not have a lot of experience with threads in golang to be honest xD

**Comment by Giulio2002** [comment]
now it uses `*atomic.Value`

File: ethdb/privateapi/ethbackend.go [hunk 3]
```diff
@@ -270,6 +289,80 @@ func (s *EthBackendServer) EngineGetPayloadV1(ctx context.Context, req *remote.E
...
+	}
+	// Check if parent equate to the head
+	parent := gointerfaces.ConvertH256ToHash(req.Forkchoice.HeadBlockHash)
+	tx, err := s.db.BeginRo(ctx)
+	if err != nil {
+		return nil, err
+	}
+	block := s.assembledBlock.Load().(*types.Block)
+
+	if parent != rawdb.ReadHeadBlockHash(tx) || block.Header().ParentHash != parent || *s.waitingForBeaconChain == 0 {
```
Comment (yperbasis) [inline ethdb/privateapi/ethbackend.go:310]:
```suggestion
	if parent != rawdb.ReadHeadBlockHash(tx) || block.Header().ParentHash != parent || atomic.LoadUint32(s.waitingForBeaconChain) == 0 {
```

File: ethdb/privateapi/ethbackend.go [hunk 4]
```diff
@@ -270,6 +289,80 @@ func (s *EthBackendServer) EngineGetPayloadV1(ctx context.Context, req *remote.E
...
 }
 
+// EngineGetPayloadV1, retrieves previously assembled payload (Validators only)
+func (s *EthBackendServer) EngineForkChoiceUpdatedV1(ctx context.Context, req *remote.EngineForkChoiceUpdatedRequest) (*remote.EngineForkChoiceUpdatedReply, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.config.TerminalTotalDifficulty == nil {
+		return nil, fmt.Errorf("not a proof-of-stake chain")
+	}
+	if s.assembledBlock.Load() == nil {
```
Comment (yperbasis) [inline ethdb/privateapi/ethbackend.go:299]: `payloadId` in the request may be null, in which case we shouldn't care about mining availability

File: cmd/rpcdaemon/commands/engine_api.go [hunk 5]
```diff
@@ -37,6 +39,13 @@ type ExecutionPayload struct {
 	Transactions  []hexutil.Bytes `json:"transactions"  gencodec:"required"`
 }
 
+// PayloadAttributes represent the attributes required to start assembling a payload
+type ForkChoiceState struct {
+	HeadHash           common.Hash `json:"headBlockHash"             gencodec:"required"`
+	SafeBlockHash      common.Hash `json:"safeBlockHash"             gencodec:"required"`
+	FizalizedBlockHash common.Hash `json:"finalizedBlockHash"        gencodec:"required"`
```
Comment (yperbasis) [inline cmd/rpcdaemon/commands/engine_api.go:46]: typo: FizalizedBlockHash

File: eth/stagedsync/stage_headers.go [hunk 6]
```diff
@@ -135,12 +124,26 @@ func HeadersPOS(
...
+	if tx != nil {
+		if err := tx.Commit(); err != nil {
+			return err
+		}
+	}
 	atomic.StoreUint32(cfg.waitingPosHeaders, 1)
 	// Waiting for the beacon chain
 	log.Info("Waiting for payloads...")
-	header := <-cfg.reverseDownloadCh
+	payloadMessage := <-cfg.reverseDownloadCh
```
Comment (yperbasis) [inline eth/stagedsync/stage_headers.go:137]: I'd move the waitingPosHeaders manipulations closer together to reduce the chance of an incorrect value. I.e. something like
```
...
log.Info("Waiting for payloads...")
atomic.StoreUint32(cfg.waitingPosHeaders, 1)
payloadMessage := <-cfg.reverseDownloadCh
atomic.StoreUint32(cfg.waitingPosHeaders, 0)
header := payloadMessage.Header
...
```

File: eth/stagedsync/stage_headers.go [hunk 7]
```diff
@@ -172,14 +177,14 @@ func HeadersPOS(
 		return err
 	}
 	if parent != nil && parent.Hash() == header.ParentHash {
-		if err := cfg.hd.VerifyHeader(&header); err != nil {
+		/*if err := cfg.hd.VerifyHeader(header); err != nil {
```
Comment (yperbasis) [inline eth/stagedsync/stage_headers.go:180]: What was the problem with header verification?

File: eth/stagedsync/stage_mining_create_block.go [hunk 8]
```diff
@@ -29,6 +29,7 @@ type MiningBlock struct {
 	Uncles   []*types.Header
 	Txs      []types.Transaction
 	Receipts types.Receipts
+	POS      bool
```
Comment (yperbasis) [inline eth/stagedsync/stage_mining_create_block.go:32]: Unused field!

File: eth/stagedsync/stage_headers.go (hunk 7 above)
Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:180]: GasLimit was wrongly detected as faulty, (aka too much change), i am not sure if this is a problem regarding the engine or regarding mergemock(this only happens when running mergemock), but until we connect to kintsugi and see for ourselves i think this can be commented out

Comment (yperbasis) [reply eth/stagedsync/stage_headers.go:179]: We should not to comment out header verification but rather investigate any issues with it and fix them. We can do that later, and for now I don't see a point in merging disabling header verification. Do you have instructions on how to test Erigon with mergemock handy? (So that I could reproduce the gas limit issue.)

Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:179]: I will write some notes then on the procedure and send them to you... i will uncomment this in the meanwhile...

File: eth/stagedsync/stage_mining_finish.go [hunk 9]
```diff
@@ -46,6 +50,15 @@ func SpawnMiningFinishStage(s *StageState, tx kv.RwTx, cfg MiningFinishCfg, quit
 	block := types.NewBlock(current.Header, current.Txs, current.Uncles, current.Receipts)
```
Comment (yperbasis) [inline eth/stagedsync/stage_mining_finish.go:50]: Uncles are prohibited after the Merge. I'm thinking that we probably should remove uncles altogether from our mining. It's unlikely that anybody will use Erigon for mining before the Merge (and even if they do, it's a small loss not to include uncles).

File: ethdb/privateapi/ethbackend.go [hunk 10]
```diff
@@ -270,6 +289,80 @@ func (s *EthBackendServer) EngineGetPayloadV1(ctx context.Context, req *remote.E
...
+
+		err := rlp.Encode(buf, tx)
+		if err != nil {
+			return nil, fmt.Errorf("broken tx rlp: %w", err)
+		}
+		encodedTransactions = append(encodedTransactions, common.CopyBytes(buf.Bytes()))
+	}
+	// Compute the correct block hash, by setting up the right header
+	header := block.Header()
+	header.Difficulty = serenity.SerenityDifficulty
```
Comment (yperbasis) [inline ethdb/privateapi/ethbackend.go:335]: stage_mining_create_block calls engine.Prepare, which in serenity consensus engine sets difficulty & nonce, so we can omit them here.

File: ethdb/privateapi/ethbackend.go [hunk 11]
```diff
@@ -270,6 +289,80 @@ func (s *EthBackendServer) EngineGetPayloadV1(ctx context.Context, req *remote.E
...
+	header.Nonce = serenity.SerenityNonce
+	header.MixDigest = gointerfaces.ConvertH256ToHash(req.Prepare.Random)
+	header.Time = req.Prepare.Timestamp
+	header.Coinbase = gointerfaces.ConvertH160toAddress(req.Prepare.FeeRecipient)
```
Comment (yperbasis) [inline ethdb/privateapi/ethbackend.go:339]: This is actually problematic. In the assembled block all transaction fees went to the original coinbase (not sure how that one is set). We cannot simply replace coinbase post `stage_mining_exec`. For now a quick hack would be to **not** set `header.Coinbase` to the FeeRecipient (and [the spec](https://github.com/ethereum/execution-apis/blob/v1.0.0-alpha.5/src/engine/specification.md#payloadattributesv1) actually allows us to do it; there FeeRecipient is only a _suggested_ value).

Going forward, however, we need to reverse the order and start assembling the block with the right fee recipient after we have received `engine_forkchoiceUpdatedV1`, not beforehand.

**Review by yperbasis — CHANGES_REQUESTED** [review]

File: ethdb/privateapi/ethbackend.go (hunk 11 above)
Comment (Giulio2002) [reply ethdb/privateapi/ethbackend.go:337]: oh okay, that actually could be fixed if we run the mining stages directly from ethbackend, there is then also a bug now that i look at it when calling the timestamp and coinbase opcode they would be unmatching

Comment (Giulio2002) [reply ethdb/privateapi/ethbackend.go:337]: let me convert this to draft again, this needs to be fixed

Comment (Giulio2002) [reply ethdb/privateapi/ethbackend.go:337]: we also are not using correct gasLimit then so that might be an indication why verfication fails

Comment (Giulio2002) [reply ethdb/privateapi/ethbackend.go]: now it is fixed.

**Comment by Giulio2002** [comment]
Now block assembling is done synchronously

File: .gitignore (hunk 1 above)
Comment (tbcd) [reply .gitignore]: > the `.gitignore` will be cleaned prior to merging, i put it like that so that i dont have to make a new one every time

The local file `.git/info/exclude` can be handy for this.

File: eth/stagedsync/stage_mining_create_block.go [hunk 12]
```diff
@@ -34,10 +34,17 @@ type MiningBlock struct {
 	RemoteTxs types.TransactionsStream
 }
 
+// In case we are proposing during proof-of-stake and we are supplied with header fields already
+type PresetHeaderFields struct {
+	Timestamp uint64
+	Random    common.Hash
```
Comment (yperbasis) [inline eth/stagedsync/stage_mining_create_block.go:40]: Let's add fee recipient as well.

File: eth/stagedsync/stage_headers.go [hunk 13]
```diff
@@ -95,16 +95,6 @@ func SpawnStageHeaders(
 	initialCycle bool,
 	test bool, // Set to true in tests, allows the stage to fail rather than wait indefinitely
 ) error {
-	useExternalTx := tx != nil
```
Comment (yperbasis) [inline eth/stagedsync/stage_headers.go:98]: Why did you remove this code?

File: eth/stagedsync/stage_headers.go [hunk 14]
```diff
@@ -135,13 +124,26 @@ func HeadersPOS(
 	cfg HeadersCfg,
 	initialCycle bool,
 	test bool, // Set to true in tests, allows the stage to fail rather than wait indefinitely
-	useExternalTx bool,
 ) error {
-	atomic.StoreUint32(cfg.waitingPosHeaders, 1)
+	// We need to have another write transaction for the miner in case we want to validate.
```
Comment (yperbasis) [inline eth/stagedsync/stage_headers.go:128]: Not sure I'm following. Why do you commit the transaction?

Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:128]: we are closing it, in case we receive the request to mine. Mining stages uses a write transaction so we would stall when trying to propose new blocks

File: eth/stagedsync/stage_headers.go (hunk 13 above)
Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:98]: no transactions in HeadersPOS will be external

Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:98]: we may need a write transaction for proposing

File: eth/stagedsync/stage_mining_create_block.go [hunk 15]
```diff
@@ -288,7 +302,13 @@ func SpawnMiningCreateBlockStage(s *StageState, tx kv.RwTx, cfg MiningCreateBloc
 	}
 
 	current.Header = header
-	current.Uncles = makeUncles(env.uncles)
+	if isTrans {
+		current.Header.Time = cfg.presetFields.Timestamp
+		current.Header.MixDigest = cfg.presetFields.Random
+		current.Uncles = nil
+	} else {
+		current.Uncles = makeUncles(env.uncles)
```
Comment (enriavil1) [inline eth/stagedsync/stage_mining_create_block.go:311]: Do we need this if we are just going to get rid of uncles entirely for POS?

File: eth/stagedsync/stage_headers.go (hunk 13 above)
Comment (yperbasis) [reply eth/stagedsync/stage_headers.go:98]: Whether `tx` is nil or not depends on `canRunCycleInOneTransaction` in `StageLoopStep`, doesn't it? So `SpawnStageHeaders` may be called with `tx == nil`, right? Then `rawdb.Transitioned(tx, ...)`at line 106 will fail.

Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go]: Oh, okay fair,might have removed it by mistake. fixed

File: eth/stagedsync/stage_mining_create_block.go (hunk 15 above)
Comment (Giulio2002) [reply eth/stagedsync/stage_mining_create_block.go:311]: we just discard them completely for now, in the future we will remove all Uncle logic

File: eth/stagedsync/stage_headers.go [hunk 16]
```diff
@@ -324,9 +341,6 @@ func HeadersPOW(
 	var headerProgress uint64
 	var err error
 
-	if !useExternalTx {
```
Comment (yperbasis) [inline eth/stagedsync/stage_headers.go:327]: Why did you remove this?

File: eth/stagedsync/stage_headers.go [hunk 17]
```diff
@@ -188,18 +201,25 @@ func HeadersPOS(
 			LatestValidHash: headerHash,
 		}
 
-		if err := headerInserter.FeedHeaderPoS(tx, &header, headerHash); err != nil {
+		if err := headerInserter.FeedHeaderPoS(tx, header, headerHash); err != nil {
+			return err
+		}
+		// We can insert raw bodies immediately and skip bodies
+		if err := rawdb.WriteRawBody(tx, headerHash, headerNumber, payloadMessage.Body); err != nil {
```
Comment (yperbasis) [inline eth/stagedsync/stage_headers.go:208]: And we also need to write the block hash.

File: eth/stagedsync/stage_headers.go (hunk 14 above)
Comment (yperbasis) [reply eth/stagedsync/stage_headers.go:137]: Hmm. To gut feeling is that the transaction commitment here might be problematic. I don't understand the current interplay between waiting for PoS blocks and the mining. Probably need to pick up @AskAlexSharov brain for that.

File: eth/stagedsync/stage_headers.go (hunk 17 above)
Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:208]: block hash is another stage not in bodies

Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:208]: there is a blockhashes stage for this

Comment (yperbasis) [reply eth/stagedsync/stage_headers.go:208]: Ah, OK. It's a bit inconsistent though: you let the 2nd stage (BlockHashes) to run its natural course, while you save the block and manually set the progress on the 3rd stage (Bodies). Better at least put a comment in the code about this inconsistency.

File: eth/stagedsync/stage_headers.go (hunk 16 above)
Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:327]: because it is unneded, it should have already removed it with `SpawnHeadersStage` but forgot

File: eth/stagedsync/stage_headers.go (hunk 14 above)
Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:137]: I do not think it is problematic, if any error occur we rollback them, so if the tx has not been killed and has been kept by another stage it should not give problems. if any stage does not rollback on error then the issue is not in the scope of stage headers. I can explain the interplay, mining stages uses a write transaction and block proposing happens only if we are waiting the beacon chain, thus when mining is requested the stagedsync must be in stage headers waiting for payloads. now if we did not close it prior waiting, the mining stage will never start unless we close the write transaction in stage headers. because you cannot have more than 2 write transactions on MDBX.

File: eth/stagedsync/stage_headers.go (hunk 17 above)
Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:209]: added a TODO

File: eth/stagedsync/stage_headers.go (hunk 14 above)
Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:137]: Note also that external transactions are usually used for unit tests and not for production

Comment (AskAlexSharov) [reply eth/stagedsync/stage_headers.go:137]: "external transactions are usually used for unit tests and not for production" - not true. 2 setups used in production: 
- if jump < 8K blocks - then whole staged sync step run inside 1 write transaction: means have only 1 fsync after new block processed, instead of fsync on each stage.
- if jump > 8K blocks (genesis sync, and long downtimes) - no external transaction started and stages can save partial progress as they want.

Comment (AskAlexSharov) [reply eth/stagedsync/stage_headers.go:137]: About Mining and 2 write txs: 
Now, next code used to stop staged sync and release write tx - so, mining can start own write tx: 
```
case <-cfg.hd.SkipCycleHack:
			break Loop
```

Comment (AskAlexSharov) [reply eth/stagedsync/stage_headers.go:137]: In my understanding: until "beacon validation done" we can't mine - because we don't know on which fork to mine (if beacon will say that block is valid - need mine on new block, if beacon say that block isn't valid - then on old block).

"waiting for beacon" - what is maximum time we can wait? As I understand it's synchronous call and we must just wait?

For example "waiting for p2p messages" is async - and implemented as `case <-cfg.hd.DeliveryNotify:`, can be interrupted by `case <-cfg.hd.SkipCycleHack:`. Then on next sync cycle all non-processed async messages will be handled.

Comment (yperbasis) [reply eth/stagedsync/stage_headers.go:137]: Yeah, we can probably use `SkipCycleHack` in `HeadersPOS` alongside `reverseDownloadCh` (listen on both). If something arrives to `SkipCycleHack`, then `HeadersPOS` should return so that the stage sync loop proceeds to the next stages, all the way to the mining.

Comment (yperbasis) [reply eth/stagedsync/stage_headers.go:137]: And when the Beacon node sends us a [engine_forkchoiceupdatedv1](https://github.com/ethereum/execution-apis/blob/v1.0.0-alpha.5/src/engine/specification.md#engine_forkchoiceupdatedv1) request with a payload, ethbackend should send a message to `SkipCycleHack` to interrupt the wait in the header stage.

Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:137]: okay, so to answer all the points:
* Beacon chain calls are synchronous
* if external transactions are used in production then ending stage headers simply wont't be enough
* I noticed that there is an actual benefit of doing how i did it as of now, before we mined block every 3 seconds on a goroutine constantly, now we can do the same without having a goroutine constantly on and  only when requested.

Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:137]: I think it makes more sense logically to do it like this, also tests are passing so it should not break any other component

Comment (Giulio2002) [reply eth/stagedsync/stage_headers.go:137]: yes, must just wait until we timeout

File: eth/stagedsync/stage_mining_create_block.go [hunk 18]
```diff
@@ -34,10 +34,18 @@ type MiningBlock struct {
...
+	Timestamp             uint64
+	Random                common.Hash
+	SuggestedFeeRecipient common.Address
+}
+
 type MiningState struct {
 	MiningConfig    *params.MiningConfig
 	PendingResultCh chan *types.Block
 	MiningResultCh  chan *types.Block
+	ProposeBlockCh  chan *types.Block
```
Comment (yperbasis) [inline eth/stagedsync/stage_mining_create_block.go:48]: Why not reuse `MiningResultCh`? If the idea is to have a separate channel for PoS blocks, let's call it `PoSMiningResultCh` or something.

Comment (Giulio2002) [reply eth/stagedsync/stage_mining_create_block.go:48]: because it might be used for something else, also i would not name it Mining because we are not mining, we are proposing

Comment (Giulio2002) [reply eth/stagedsync/stage_mining_create_block.go:48]: which is a similar, yet fundamentally different process

File: ethdb/privateapi/ethbackend.go [hunk 19]
```diff
@@ -270,6 +291,74 @@ func (s *EthBackendServer) EngineGetPayloadV1(ctx context.Context, req *remote.E
 	return nil, fmt.Errorf("unknown payload")
 }
 
+// EngineGetPayloadV1, retrieves previously assembled payload (Validators only)
```
Comment (yperbasis) [inline ethdb/privateapi/ethbackend.go:294]: The comment doesn't match the function

Comment (Giulio2002) [reply ethdb/privateapi/ethbackend.go:294]: it does, getPayload returns a previously assembled aka. proposed block

Comment (Giulio2002) [reply ethdb/privateapi/ethbackend.go:294]: oh wait, my bad

File: eth/backend.go [hunk 20]
```diff
@@ -404,9 +404,29 @@ func New(stack *node.Node, config *ethconfig.Config, logger log.Logger) (*Ethere
 	if casted, ok := backend.engine.(*ethash.Ethash); ok {
 		ethashApi = casted.APIs(nil)[1].Service.(*ethash.API)
 	}
-	atomic.StoreUint32(&backend.waitingForPOSHeaders, 0)
+	atomic.StoreUint32(&backend.waitingForBeaconChain, 0)
+	startAssembleFunc := func(timestamp uint64, random common.Hash, suggestedFeeRecipient common.Address) (types.Block, error) {
```
Comment (yperbasis) [inline eth/backend.go:408]: We should not perform complete mining here, only to initiate it. Per the spec: "if payloadAttributes is not null and the client is not SYNCING, and MUST **begin** a payload build process building on top of forkchoiceState.headBlockHash".  And the build process should start concurrently and last for up to 12 seconds. See https://github.com/ethereum/execution-apis/blob/v1.0.0-alpha.5/src/engine/specification.md#payload-build-process

Currently `startAssembleFunc` confusingly not only starts the build process, but waits for it to finish.

File: eth/stagedsync/stage_headers.go (hunk 14 above)
Comment (yperbasis) [reply eth/stagedsync/stage_headers.go:137]: The thing is that Beacon mining ("payload build process") is supposed to be asynchronous. engine_forkchoiceUpdatedV1 with payload should only initiate the mining, not to wait for its completion. So IMHO more work is still required.

But this PR is too big already. In order to make progress, I suggest the following:
1. Leave only a rudimentary block build process in this PR, without any transactions (probably w/o correct state root either). Then we continue the work on proper PoS mining with transactions and everything in a subsequent PR.
2. Restore the `useExternalTx` logic in `HeaderPoS` and don't do this `tx.Commit` hack.

File: eth/stagedsync/stage_mining_create_block.go (hunk 18 above)
Comment (yperbasis) [reply eth/stagedsync/stage_mining_create_block.go:48]: OK, fair enough, but ProposeBlockCh still lives inside MiningState :) To my mind it's fine to call PoS payload building mining, but it's not that important.

File: ethdb/privateapi/ethbackend.go [hunk 21]
```diff
@@ -196,9 +206,11 @@ func (s *EthBackendServer) EngineExecutePayloadV1(ctx context.Context, req *type
 	}
 
 	blockHash := gointerfaces.ConvertH256ToHash(req.BlockHash)
-
+	// Discard all previous prepared payloads if another block was proposed
+	s.nextPayloadId = 0
```
Comment (yperbasis) [inline ethdb/privateapi/ethbackend.go:210]: Why is nextPayloadId set to zero every time EngineExecutePayloadV1 is called?

Comment (Giulio2002) [reply ethdb/privateapi/ethbackend.go:210]: if a new block is executed, it means that we can discard assembled block based on the previous state, nextPayloadId is set to 0 so that we do not risk overflows and what not, just seemed a good thing to do

Comment (yperbasis) [reply ethdb/privateapi/ethbackend.go:210]: Resetting payload ID might be confusing to the Consensus Layer. It'll also complicate tracking and debugging of different payloads. If we manage to overflow 2^64, then something went very wrong. Beacon epoch is about 12 sec, and Engine API requests should come with roughly that frequency. Thus we have about 7*10^12 years as a safety margin :)

Comment (Giulio2002) [reply ethdb/privateapi/ethbackend.go]: that is fine to remove, it was more of personal preference, i do not think it gives problems to beacon node because if you turned off any go-ethereum then it would not remember which payloadId goes to which payload

File: ethdb/privateapi/ethbackend.go [hunk 22]
```diff
@@ -270,6 +285,56 @@ func (s *EthBackendServer) EngineGetPayloadV1(ctx context.Context, req *remote.E
...
+	if err != nil {
+		return nil, err
+	}
+
+	headHeader, err := rawdb.ReadHeaderByHash(tx, parent)
+	if err != nil {
+		return nil, err
+	}
+
+	if atomic.LoadUint32(s.waitingForBeaconChain) == 0 || headHeader == nil {
```
Comment (yperbasis) [inline ethdb/privateapi/ethbackend.go:307]: The previous check `parent != rawdb.ReadHeadBlockHash(tx)` is better since we might have the header in the DB but be on a different fork.

Comment (Giulio2002) [reply ethdb/privateapi/ethbackend.go:307]: we are reading the header by hash, which means it looks into blockhashes which does not do only canonical headers

File: ethdb/privateapi/ethbackend.go [hunk 23]
```diff
@@ -270,6 +285,56 @@ func (s *EthBackendServer) EngineGetPayloadV1(ctx context.Context, req *remote.E
...
+		ExtraData:     headHeader.Extra,
+		BaseFeePerGas: &types2.H256{},
+		BlockHash:     gointerfaces.ConvertHashToH256(headHeader.Hash()),
+		Transactions:  [][]byte{},
+	}
+	// successfully assembled the payload and assinged the correct id
+	defer func() { s.nextPayloadId++ }()
+	return &remote.EngineForkChoiceUpdatedReply{
+		Status:    "SUCCESS",
+		PayloadId: s.nextPayloadId,
```
Comment (yperbasis) [inline ethdb/privateapi/ethbackend.go:334]: Suggestion: rename `nextPayloadId` to `payloadId` because it's used as the ID of the current payload, not the next one.

Comment (Giulio2002) [reply ethdb/privateapi/ethbackend.go]: done

File: ethdb/privateapi/ethbackend.go [hunk 24]
```diff
@@ -270,6 +285,56 @@ func (s *EthBackendServer) EngineGetPayloadV1(ctx context.Context, req *remote.E
...
+		Coinbase:      req.Prepare.FeeRecipient,
+		Timestamp:     req.Prepare.Timestamp,
+		Random:        req.Prepare.Random,
+		StateRoot:     gointerfaces.ConvertHashToH256(headHeader.Root),
+		ReceiptRoot:   gointerfaces.ConvertHashToH256(types.EmptyRootHash),
+		LogsBloom:     &types2.H2048{},
+		GasLimit:      headHeader.GasLimit,
+		GasUsed:       0,
+		BlockNumber:   headHeader.Number.Uint64() + 1,
+		ExtraData:     headHeader.Extra,
```
Comment (yperbasis) [inline ethdb/privateapi/ethbackend.go:325]: Let's set ExtraData to empty

File: ethdb/privateapi/ethbackend.go [hunk 25]
```diff
@@ -270,6 +285,56 @@ func (s *EthBackendServer) EngineGetPayloadV1(ctx context.Context, req *remote.E
...
+		return nil, err
+	}
+
+	if atomic.LoadUint32(s.waitingForBeaconChain) == 0 || headHeader == nil {
+		return &remote.EngineForkChoiceUpdatedReply{
+			Status: string(Syncing),
+		}, nil
+	}
+
+	// Hash is incorrect because mining archittecture has yet to be implemented
```
Comment (yperbasis) [inline ethdb/privateapi/ethbackend.go:313]: Hash can be calculated from the header. Theoretically we have enough data to construct a valid empty block. Since there are no transactions and no block rewards (after the Merge), state root doesn't change. We also would need to calculate the base fee per EIP-1559 rules.
On the other hand, perhaps it's not worth it to bother since we're going to add proper PoS mining with transactions in a subsequent PR.

**Review by yperbasis — APPROVED** [review]

File: ethdb/privateapi/ethbackend.go (hunk 22 above)
Comment (yperbasis) [reply ethdb/privateapi/ethbackend.go:307]: Yeah, that's my point. When our latest canonical header doesn't match headBlockHash, we should return "SYNCING" status and initiate a re-org.

File: ethdb/privateapi/ethbackend.go (hunk 25 above)
Comment (Giulio2002) [reply ethdb/privateapi/ethbackend.go:313]: yup, that was my point

File: ethdb/privateapi/ethbackend.go (hunk 22 above)
Comment (Giulio2002) [reply ethdb/privateapi/ethbackend.go:307]: enrique will add it
